src/macaron/repo_finder/commit_finder.py

```diff
@@ -0,0 +1,397 @@
+
+    if not matched_tags:
+        logger.debug("No tags matched for %s", str(purl))
+        return "", ""
+
+    if len(matched_tags) > 1:
+        logger.debug("Tags found for %s: %s", str(purl), len(matched_tags))
+        logger.debug("Best match: %s", matched_tags[0])
+        logger.debug("Up to 5 others: %s", matched_tags[1:6])
+
+    tag = matched_tags[0]
+    tag_name = str(tag)
+    branches = git_obj.get_commit_from_tag(tag_name).branches
+    logger.debug("Branches: %s", branches)
+
+    if not branches:
+        logger.debug("No branch associated with tag (commit): %s (%s)", tag_name, tag.commit.hexsha)
+        return "", ""
+
+    branch_name = ""
+    for branch in branches:
+        # Ensure the detached head branch is not picked up.
+        if "(HEAD detached at" not in branch:
+            branch_name = branch
+            break
+
+    if not branch_name:
+        logger.debug("No valid branch associated with tag (commit): %s (%s)", tag_name, tag.commit.hexsha)
+        return "", ""
+
+    logger.debug(
+        "Found tag %s with commit %s of branch %s for artifact version %s@%s",
+        tag,
+        tag.commit.hexsha,
+        branch_name,
+        purl.name,
+        purl.version,
+    )
+    return branch_name, tag.commit.hexsha
+
+
+def _build_version_pattern(version: str) -> tuple[Pattern, list[str], bool]:
+    """Build a version pattern to match the passed version string.
+
+    Parameters
+    ----------
+    version: str
+        The version string.
+
+    Returns
+    -------
+    tuple[Pattern, list[str], bool]
+        The tuple of the regex pattern that will match the version, the list of version parts that were extracted, and
+        whether the version string has a non-numeric suffix.
+
+    """
+    # The version is split on non-alphanumeric characters to separate the version parts from the non-version parts.
+    # e.g. 1.2.3-DEV -> [1, 2, 3, DEV]
+    split = split_pattern.split(version)
+    logger.debug("Split version: %s", split)
+    if not split:
+        # If the version string contains no separators use it as is.
+        split = [version]
+
+    this_version_pattern = ""
+    parts = []
+    numeric_parts = 0
+    has_non_numeric_suffix = False
+    for count, part in enumerate(split):
+        # Validate the split part by checking it is only comprised of alphanumeric characters.
+        valid = validation_pattern.match(part)
+        if not valid:
+            continue
+        parts.append(part)
+
+        # Detect versions that end with a zero, so the zero can be made optional.
+        has_trailing_zero = len(split) > 2 and count == len(split) - 1 and part == "0"
+        numeric_only = numeric_only_pattern.match(str(part))
+
+        if not has_non_numeric_suffix and not numeric_only:
+            # A non-numeric part enables the flag for treating this and all remaining parts as version suffix parts.
+            has_non_numeric_suffix = True
```

Comment:
Can you give some examples of version strings that should return `has_non_numeric_suffix` to be `True`? I am unsure about the meaning of it.
From my testing, it returns `True` for the version string `v0.1.0`, which I am not sure is as intended.

Reply:
If any part of a version string contains a non-numeric character then this flag will become `True`.
E.g. `1.1A.2` -> `True`, `3.5.RELEASE` -> `True`
Your example above would cause it to be true, but only if you include the `v`. Typically, it is only tags that start with an alphabetic character like that.

Reply:
Hmm, I checked the [PURL type definitions](https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst) page and it does seem that most version strings do not have a `v` at the start.

Reply:
It makes sense to leave some comments in the code to describe what a `has_non_numeric_suffix` is (some examples would be nice), and also to describe its effect on the regex pattern being built.

Reply:
I want to confirm if I understand this right. Let's take one of the examples you gave:
```
1.1A.2
```
Here the regex generated will be "1(...)?", since the `1A` part switches `has_non_numeric_suffix` to `True`. I assume your assumption here is: parts that are not all-numeric are likely insignificant and they can be matched optionally.

Reply:
FYI the version string `1.1A.2` results in `has_non_numeric_suffix` being `False`.
I added the following `print` statements to figure out what is going on:
```diff
         # Detect versions that end with a zero, so the zero can be made optional.
         has_trailing_zero = len(split) > 2 and count == len(split) - 1 and part == "0"
         numeric_only = numeric_only_pattern.match(str(part))
+        print(f"{part=}")
+        print(f"{numeric_only=}")
```
and called `_build_version_pattern("1.1A.2")`. This printed
```
part='1'
numeric_only=<re.Match object; span=(0, 1), match='1'>
part='1A'
numeric_only=<re.Match object; span=(0, 1), match='1'>
part='2'
numeric_only=<re.Match object; span=(0, 1), match='2'>
```
which is wrong, because `1A` should not be matched as being "numeric only".
I think the issue here is with the `numeric_only_pattern` regex. Currently, it is
```python3
numeric_only_pattern = re.compile("[0-9]+")
```
This pattern still matches against `1A` because it is not exhaustive. What we need here is
```diff
-numeric_only_pattern = re.compile("[0-9]+")
+numeric_only_pattern = re.compile("^[0-9]+$")
```
After this fix, the result is now as expected.
I suspect some other regex patterns also need to be changed, one of which is `validation_pattern`.